services/gundeck/test/integration/API.hs

```diff
@@ -1007,6 +1023,22 @@ connectUsersAndDevicesWithSendingClients ca uidsAndConnIds = do
   (\(uid, conns) -> wsAssertPresences uid (length conns)) `mapM_` uidsAndConnIds
   pure chs
 
+connectUsersAndDevicesWithSendingClientsRaw ::
+  HasCallStack =>
+  CannonR ->
+  [(UserId, [ConnId])] ->
+  TestM [(UserId, [(TChan WS.Message, TChan ByteString)])]
+connectUsersAndDevicesWithSendingClientsRaw ca uidsAndConnIds = do
+  chs <- forM uidsAndConnIds $ \(uid, conns) ->
+    (uid,) <$> do
+      forM conns $ \conn -> do
+        chread <- liftIO $ atomically newTChan
+        chwrite <- liftIO $ atomically newTChan
+        _ <- wsRun ca uid conn (wsReaderWriterPing chread chwrite)
+        pure (chread, chwrite)
+  (\(uid, conns) -> wsAssertPresences uid (length conns)) `mapM_` uidsAndConnIds
```

Comment:
I'm wondering if this wouldn't become a bit simpler, if you would move `wsAssertPresences uid (length conns)` into the scope of the `forM` expression above :thinking:

Reply:
I refactored this a little in https://github.com/wireapp/wire-server/pull/3047/commits/61791b5ef670a129fdc89b5c031685fe1afc2674